core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java

```diff
@@ -230,13 +233,37 @@ public void clearLockedKeys() {
       lockedKeys = null;
    }
 
-   private boolean hasLockOrIsLockBackup(Object key) {
-      //stopgap fix for ISPN-2728. The real fix would be to synchronize this with the intrinsic lock.
+   @Override
+   public boolean containsLockOrBackupLock(Object key) {
       Set<Object> lockedKeysCopy = lockedKeys;
       Set<Object> backupKeyLocksCopy = backupKeyLocks;
       return (lockedKeysCopy != null && lockedKeysCopy.contains(key)) || (backupKeyLocksCopy != null && backupKeyLocksCopy.contains(key));
    }
 
+   @Override
+   public boolean containsAnyLockOrBackupLock(Stream<Object> keys) {
+      return findFirstLockedOrBackupLocked(keys) != null;
+   }
+
+   @Override
+   public Object findFirstLockedOrBackupLocked(Stream<Object> keys) {
+      Set<Object> lockedKeysCopy = lockedKeys;
+      Set<Object> backupKeyLocksCopy = backupKeyLocks;
+      if (lockedKeysCopy != null && backupKeyLocksCopy != null) {
+         return keys.filter(key -> lockedKeysCopy.contains(key) || backupKeyLocksCopy.contains(key)).findAny().orElse(null);
+      } else if (lockedKeysCopy != null) {
+         return keys.filter(lockedKeysCopy::contains).findAny().orElse(null);
+      } else if (backupKeyLocksCopy != null) {
+         return keys.filter(backupKeyLocksCopy::contains).findAny().orElse(null);
+      }
+      return null;
+   }
```

Comment:
I can understand that yes.  Also when using iteration, if using an array list we should use for int loops instead as they are also 16% faster.

Although we need to be cognizant of what the alternatives are more than just saying 1 thing is bad imo.  For example iterating over the collection multiple times or creating a new collection could be worse then just using a stream.